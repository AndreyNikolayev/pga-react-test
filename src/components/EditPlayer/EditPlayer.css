.EditPlayer {
    margin-top:10px;
}

.EditPlayer input {
    max-width: 300px;
}

.EditPlayer input[type="number"] {
    max-width: 60px;
}

.EditPlayer .cancel-btn {
    margin-right: 20px;
}