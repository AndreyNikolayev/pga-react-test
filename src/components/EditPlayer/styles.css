.edit-player {
    margin-top:10px;
}

.edit-player input {
    max-width: 300px;
}

.edit-player input[type="number"] {
    max-width: 60px;
}

.edit-player .cancel-btn {
    margin-right: 20px;
}