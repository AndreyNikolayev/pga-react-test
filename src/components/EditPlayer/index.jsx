import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import './styles.css';
import PropTypes from 'prop-types';

export default class EditPlayer extends Component {
  constructor(props) {
    super(props)

    this.onChangeInput = this.onChangeInput.bind(this)
    this.onSubmit = this.onSubmit.bind(this)

    this.state = {
      playerId: this.props.match.params.id,
      firstName: this.composeInput('firstName'),
      lastName: this.composeInput('lastName'),
      score: this.composeInput('score')
    }
  }

  componentWillUpdate(nextProps, prevState) {
    if (nextProps.match.params.id === prevState.playerId) {
      return;
    }
    this.setState({
      playerId: this.props.match.params.id,
      firstName: this.composeInput('firstName'),
      lastName: this.composeInput('lastName'),
      score: this.composeInput('score')
    })
  }

  composeInput(inputName) {
    const inputValue = this.props.player ? this.props.player[inputName] : ''
    return {
      value: inputValue,
      isTouched: false,
      errorMessage: this.validateInput(inputName, inputValue)
    }
  }

  onTouchedInput(inputName) {
    this.setState({
      [inputName]: {
        ...this.state[inputName],
        isTouched: true
      }
    })
  }

  onChangeInput(inputName, e) {
    this.setState({
      [inputName]: {
        ...this.state[inputName],
        isTouched: false,
        value: e.target.value,
        errorMessage: this.validateInput(inputName, e.target.value)
      }
    })
  }

  validateInput(inputName, value) {
    if (value === '') {
      return 'Field is required'
    }
    if (inputName === 'score') {
      if (!/^\d+$/.test(value)) {
        return 'Only digits are allowed';
      }
      const numberValue = +value;
      if (numberValue > 100) {
        return 'Value must be between 0 and 100';
      }
    }
    return '';
  }

  async onSubmit(e) {
    e.preventDefault();
    if (this.state.firstName.errorMessage || this.state.lastName.errorMessage || this.state.score.errorMessage) {
      this.setState({
        firstName: {
          ...this.state.firstName,
          isTouched: true
        },
        lastName: {
          ...this.state.lastName,
          isTouched: true
        },
        score: {
          ...this.state.score,
          isTouched: true
        },
      });
      return;
    }

    const player = {
      firstName: this.state.firstName.value,
      lastName: this.state.lastName.value,
      score: this.state.score.value
    }

    this.props.handleSave(this.state.playerId, player)
  }

  get pageTitle() {
    return this.state.playerId ? 'Edit Player' : 'Add Player';
  }

  get labelClass() {
    return this.state.playerId ? "active" : ""
  }

  isShowInputError(inputName) {
    return this.state[inputName].isTouched && this.state[inputName].errorMessage
  }

  render() {
    return (
      <div className="edit-player">
        <h3>{this.pageTitle}</h3>
        <div className="row">
          <form className="s12 m6" onSubmit={this.onSubmit} noValidate>
            <div className="row">
              <div className="input-field col s12">
                <label htmlFor="firstName" className={this.labelClass}>First Name:  </label>
                <input id="firstName"
                  type="text"
                  className={"form-control " + (this.isShowInputError('firstName') ? "invalid" : "")}
                  value={this.state.firstName.value}
                  onChange={(e) => this.onChangeInput('firstName', e)}
                  onBlur={(e) => this.onTouchedInput('firstName', e)}
                />
                <span className="helper-text" data-error={this.state.firstName.errorMessage}></span>
              </div>
              <div className="input-field col s12">
                <label htmlFor="lastName" className={this.labelClass}>Last Name: </label>
                <input id="lastName"
                  type="text"
                  className={"form-control " + (this.isShowInputError('lastName') ? "invalid" : "")}
                  value={this.state.lastName.value}
                  onChange={(e) => this.onChangeInput('lastName', e)}
                  onBlur={(e) => this.onTouchedInput('lastName', e)}
                />
                <span className="helper-text" data-error={this.state.lastName.errorMessage}></span>
              </div>
              <div className="input-field col s12">
                <label htmlFor="score" className={this.labelClass}>Score: </label>
                <input id="score"
                  type="number"
                  min="0"
                  max="100"
                  className={"form-control " + (this.isShowInputError('score') ? "invalid" : "")}
                  value={this.state.score.value}
                  onChange={(e) => this.onChangeInput('score', e)}
                  onBlur={(e) => this.onTouchedInput('score', e)}
                />
                <span className="helper-text" data-error={this.state.score.errorMessage}></span>
              </div>
              <div className="col s12">
                <Link to="/" className="cancel-btn waves-effect waves-light btn">Back</Link>
                <button type="submit" className="waves-effect waves-light btn">Save</button>
              </div>
            </div>
          </form>
        </div>
      </div>
    )
  }
}

EditPlayer.propTypes = {
  player: PropTypes.object,
  save: PropTypes.func.isRequired
}
