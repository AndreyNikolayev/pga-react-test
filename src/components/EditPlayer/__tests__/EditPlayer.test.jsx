import { shallow } from 'enzyme';
import React from 'react';
import EditPlayer from '..';

describe('EditPlayer', () => {

  let wrapper;
  let mockProps;
  const saveMock = jest.fn();

  beforeEach(() => {
    mockProps = {
      handleSave: saveMock,
      playerId: '1',
      player: {
        firstName: 'First 1',
        lastName: 'Last 1',
        score: '1'
      }
    };

    wrapper = shallow(<EditPlayer {...mockProps}/>);
  })


  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('initializes inputs correctly', () => {
    const firstNameInput = wrapper.find('#firstName');
    expect(firstNameInput.props().value).toBe(mockProps.player.firstName);

    const lastNameInput = wrapper.find('#lastName');
    expect(lastNameInput.props().value).toBe(mockProps.player.lastName);

    const scoreInput = wrapper.find('#score');
    expect(scoreInput.props().value).toBe(mockProps.player.score);
  });

  it('handles input change correctly', () => {
    wrapper.find('#firstName').simulate('change', {
      target: {
        value: 'New First Name'
      }
    });
    wrapper.find('#lastName').simulate('change', {
      target: {
        value: 'New Last Name'
      }
    });
    wrapper.find('#score').simulate('change', {
      target: {
        value: '33'
      }
    });

    expect(wrapper.find('#firstName').props().value).toBe('New First Name');
    expect(wrapper.find('#lastName').props().value).toBe('New Last Name');
    expect(wrapper.find('#score').props().value).toBe('33');
  });

  it('handles blur correctly', () => {
    expect(wrapper.state('firstName').isTouched).toBeFalsy();
    expect(wrapper.state('lastName').isTouched).toBeFalsy();
    expect(wrapper.state('score').isTouched).toBeFalsy();
    

    wrapper.find('#firstName').simulate('blur');
    wrapper.find('#lastName').simulate('blur');
    wrapper.find('#score').simulate('blur');

    expect(wrapper.state('firstName').isTouched).toBeTruthy();
    expect(wrapper.state('lastName').isTouched).toBeTruthy();
    expect(wrapper.state('score').isTouched).toBeTruthy();
  });

  it('validates text inputs', () => {
    wrapper.find('#firstName').simulate('change', {
      target: {
        value: ''
      }
    });
    wrapper.find('#lastName').simulate('change', {
      target: {
        value: 'New Last Name'
      }
    });

    expect(wrapper.state('firstName').errorMessage).toBeTruthy();
    expect(wrapper.state('lastName').errorMessage).toBeFalsy();

    wrapper.find('#firstName').simulate('change', {
      target: {
        value: 'New First Name'
      }
    });
    wrapper.find('#lastName').simulate('change', {
      target: {
        value: ''
      }
    });

    expect(wrapper.state('firstName').errorMessage).toBeFalsy();
    expect(wrapper.state('lastName').errorMessage).toBeTruthy();
  });

  it('validates score', () => {
    wrapper.find('#score').simulate('change', {
      target: {
        value: ''
      }
    });
    expect(wrapper.state('score').errorMessage).toBeTruthy();

    wrapper.find('#score').simulate('change', {
      target: {
        value: '32'
      }
    });
    expect(wrapper.state('score').errorMessage).toBeFalsy();
    
    wrapper.find('#score').simulate('change', {
      target: {
        value: 'sdsg'
      }
    });
    expect(wrapper.state('score').errorMessage).toBeTruthy();

    wrapper.find('#score').simulate('change', {
      target: {
        value: '-1'
      }
    });
    expect(wrapper.state('score').errorMessage).toBeTruthy();

    wrapper.find('#score').simulate('change', {
      target: {
        value: '101'
      }
    });
    expect(wrapper.state('score').errorMessage).toBeTruthy();

    wrapper.find('#score').simulate('change', {
      target: {
        value: '100'
      }
    });
    expect(wrapper.state('score').errorMessage).toBeFalsy();

    wrapper.find('#score').simulate('change', {
      target: {
        value: '0'
      }
    });
    expect(wrapper.state('score').errorMessage).toBeFalsy();
  });

  it('handles player changed via props', () => {
    wrapper.setProps({
      playerId: '2',
      player: {
        firstName: 'First 2',
        lastName: 'Last 2',
        score: '2'
      } 
    });

   setTimeout(() => {
     wrapper.update();
     expect(wrapper.state('playerId')).toBe('2');
     expect(wrapper.state('firstName').value).toBe('First 2');
     expect(wrapper.state('lastName').value).toBe('Last 2');
     expect(wrapper.state('score').value).toBe('2');
   },0)
  });

  it('handles invalid form', () => {
    wrapper.setState({
      firstName: {
        value: '',
        errorMessage: 'Field is required',
        isTouched: true
      } 
    });

    wrapper.find('form').simulate('submit', {
      preventDefault: () => {
      }
     });

    expect(saveMock).not.toBeCalled();
    expect(wrapper.state('firstName').isTouched).toBeTruthy();
    expect(wrapper.state('lastName').isTouched).toBeTruthy();
    expect(wrapper.state('score').isTouched).toBeTruthy();
    saveMock.mockReset();
  });

  it('handles valid form', () => {

    wrapper.find('form').simulate('submit', {
      preventDefault: () => {
      }
     });

    expect(saveMock).toBeCalled();
    expect(saveMock.mock.calls[0][0]).toBe(wrapper.state('playerId'));
    expect(saveMock.mock.calls[0][1]).toEqual({
      firstName: wrapper.state('firstName').value,
      lastName:  wrapper.state('lastName').value,
      score: parseInt( wrapper.state('score').value, 10)
    });
    saveMock.mockReset();
  });
});