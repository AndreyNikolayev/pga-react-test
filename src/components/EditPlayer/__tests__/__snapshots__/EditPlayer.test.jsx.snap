// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPlayer renders correctly 1`] = `
<div
  className="edit-player"
>
  <h3>
    Edit Player
  </h3>
  <div
    className="row"
  >
    <form
      className="s12 m6"
      noValidate={true}
      onSubmit={[Function]}
    >
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <label
            className="active"
            htmlFor="firstName"
          >
            First Name:  
          </label>
          <input
            className="form-control "
            id="firstName"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="First 1"
          />
          <span
            className="helper-text"
            data-error=""
          />
        </div>
        <div
          className="input-field col s12"
        >
          <label
            className="active"
            htmlFor="lastName"
          >
            Last Name: 
          </label>
          <input
            className="form-control "
            id="lastName"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="Last 1"
          />
          <span
            className="helper-text"
            data-error=""
          />
        </div>
        <div
          className="input-field col s12"
        >
          <label
            className="active"
            htmlFor="score"
          >
            Score: 
          </label>
          <input
            className="form-control "
            id="score"
            max="100"
            min="0"
            onBlur={[Function]}
            onChange={[Function]}
            type="number"
            value="1"
          />
          <span
            className="helper-text"
            data-error=""
          />
        </div>
        <div
          className="col s12"
        >
          <Link
            className="cancel-btn waves-effect waves-light btn"
            replace={false}
            to="/"
          >
            Back
          </Link>
          <button
            className="waves-effect waves-light btn"
            type="submit"
          >
            Save
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
